src/MergeContext.js

```diff
@@ -454,6 +454,252 @@ class BranchPosition
+        let message = this._title;
+        if (this._body !== null)
+            message += '\n\n' + this._body;
+        if (this._trailer !== null)
+            message += '\n\n' + this._trailer;
+        return message;
+    }
+
+    // the future commit author in the {name, email, date} format
+    author()
+    {
+        if (this._customAuthor === null)
+            return this._defaultAuthor;
+        return {name: this._customAuthor.name, email: this._customAuthor.email, date: this._defaultAuthor.date};
+    }
+
+    // returns the position of the first non-ASCII_printable character (or -1)
+    _invalidCharacterPosition(str) {
+        const prohibitedCharacters = /[^\u{20}-\u{7e}]/u;
+        const match = prohibitedCharacters.exec(str);
+        return match ? match.index : -1;
+    }
+
+    // removes leading empty lines and trims the end
+    _trim(str) {
+        // cannot just use trim() to preserve the first line indentation (if any).
+        return str.replace(/^\s*\n+/g, '').trimEnd();
+    }
+
+    _paragraphLabel(str) {
+        if (/^\s{4}/.test(str)) // not a regular paragraph but a quotation
+            return null;
+        const rawLabel = str.match(/^\s*(\S+)\s*:/);
+        if (!rawLabel)
+            return null;
+        const label = rawLabel[1].toLowerCase();
+        return label[0].toUpperCase() + label.substring(1);
+    }
+
+    _startsWithFieldName(str) {
+        const label = this._paragraphLabel(str);
+        return (label && label.includes('-')) ? label : null;
+    }
+
+    // basic checks for an unparsed message
+    _checkRaw(message) {
+        // CRs in CRLF sequences already removed
+        // other CRs are not treated specially (and are banned)
+        const lines = message.split('\n');
+        for (let i = 0; i < lines.length; ++i) {
+            const line = lines[i];
+            const invalidPosition = this._invalidCharacterPosition(line);
+
+            if (invalidPosition !== -1)
+                throw new Error(`bad character at line ${i}, offset ${invalidPosition}`);
+
+            // allow excessively long whitespace-only lines
+            // that some copy-pasted PR descriptions may include
+            const trimmedLine = this._trim(line);
+            if (trimmedLine.length > 72)
+                throw new Error(`too long line '${trimmedLine}'`);
+        }
+    }
+
+    _parse(prDescriptionRaw) {
+        const prDescription = prDescriptionRaw.replace(/\r+\n/g, '\n');
+        this._checkRaw(prDescription);
+        const prDescriptionWithoutHeader = this._extractHeader(prDescription);
+        const prDescriptionWithoutHeaderAndTrailer = this._extractTrailer(prDescriptionWithoutHeader);
+        this._parseBody(prDescriptionWithoutHeaderAndTrailer);
+    }
+
+    // authorField is a {name, value, raw}
+    _parseAuthor(authorField) {
+        const trimmedValue = this._trim(authorField.value);
+        const cred = trimmedValue.match(/^ ([\w][^@<>,]*) <(\S+@\S+\.\S+)>$/);
```

Comment:
The leading field value space is required for the field to be parsed. It is essentially a part of the field name:value delimiter. Let's not keep that whitespace in `.value`. Let's also trim each `.value` when parsing fields so that higher-level code like this does not have to deal with beginning and ending whitespace at all. Most of these whitespace trimming changes would go into _tokenizeAll(), where the fields are parsed, but we also need to remove the extra space from the regex here:

```suggestion
        const cred = authorField.value.match(/^([\w][^@<>,]*) <(\S+@\S+\.\S+)>$/);
```

Reply:
Ok.